### DISCUSSION TITLE
Calculating player speed server-side

### DISCUSSION THREAD

**Discussion by o7q** [FabricMC/fabric-api#3976] [Mod Dev Support]
### Description

I am trying to calculate player speed on the server. I can easily do it client side but the methods I use do not work the same on the server.

I am using the `player.getX()` and `player.prevX` variables. These work perfectly fine on the client but they return the exact same value on the server.

*Server-side getX() and prevX values*
![image](https://github.com/user-attachments/assets/dd40a624-4272-4ee5-83e9-8920d6b0bf1c)

```java
double x = player.getX();
double prevX = player.prevX;

LOGGER.info("X: " + String.valueOf(x));
LOGGER.info("prevX: " + String.valueOf(prevX));
```

The function is being called from an `AttackEntityCallback` event

```java
AttackEntityCallback.EVENT.register((player, world, hand, entity, hitResult) -> {
	if (!world.isClient)
	{
		handleAttack(player);
	}
	return ActionResult.PASS;
});
```

I have been stuck on this question for the past year, worked on it on and off. I cannot find a good way to do this. I have asked around in the Discord, read the wiki, no luck yet. I presume it's something incredibly simple, I have yet to find it.

**Comment by Octol1ttle** [comment]
Use `player.getVelocity()` instead

**Reply by o7q** [reply]
`player.getVelocity()` only grabs the air velocity, it is always 0 when the player is grounded. I'd like a method that would be equivalent to taking the difference between `player.prevX/Y/Z` and `player.getPos()`. This way I don't have to worry about other mods, potion effects, or any other things that could interfere with velocity/is not included in `player.getVelocity()`.
